libraries/standard/http/src/http_client.c

```diff
@@ -1518,36 +1518,46 @@ static HTTPStatus_t sendHttpData( const HTTPTransportInterface_t * pTransport,
                                   size_t dataLen )
 {
     HTTPStatus_t returnStatus = HTTP_SUCCESS;
+    const uint8_t * pIndex = pData;
     int32_t transportStatus = 0;
+    size_t bytesRemaining = dataLen;
 
     assert( pTransport != NULL );
     assert( pTransport->send != NULL );
     assert( pData != NULL );
 
-    transportStatus = pTransport->send( pTransport->pContext,
-                                        pData,
-                                        dataLen );
-
-    if( transportStatus < 0 )
+    /* Loop until all data is sent. */
+    while( bytesRemaining > 0UL )
     {
-        LogError( ( "Failed to send HTTP data: Transport send()"
-                    " returned error: TransportStatus=%d",
-                    transportStatus ) );
-        returnStatus = HTTP_NETWORK_ERROR;
-    }
-    else if( ( size_t ) transportStatus != dataLen )
-    {
-        LogError( ( "Failed to send HTTP data: Transport layer "
-                    "did not send the required bytes: RequiredBytes=%lu"
-                    ", SentBytes=%d.",
-                    ( unsigned long ) dataLen,
-                    transportStatus ) );
-        returnStatus = HTTP_NETWORK_ERROR;
+        transportStatus = pTransport->send( pTransport->pContext,
+                                            pIndex,
+                                            bytesRemaining );
+
+        if( transportStatus > 0 )
+        {
+            bytesRemaining -= ( size_t ) transportStatus;
```

Comment:
defensive programming opportunity: if there is a bug in the `send()` routine on line 1532 which causes it to return a value greater than `bytesRemaining`, we won't exit the loop as intended with the test on line 1530 since `bytesRemaining` is unsigned.  Consider sanity checking the value of `transport_status` prior to the assignment at line 1538, e.g.:

```c
if( transportStatus > bytesRemaining )
{
    transportStatus = bytesRemaining;
    LogError( ( "Unexpected value returned from pTransport->send()...
}
```